proposals/gc/Overview.md

```diff
@@ -1 +1,728 @@
+* More complicated linking patterns might require user-space linking hooks.
+* Possibly: allow abstract type exports (encapsulation)?
+* Lots of tricky details here, mostly ignore for now...
+
+
+## Basic Functionality: Simple Aggregates
+
+* Extend the Wasm type section with new constructors for aggregate types.
+* Extend the value types with new constructors for references and interior references.
+* Aggregate types are not value types, only references to them are.
+* References are never null; nullable reference types are separate.
+
+
+### Structures
+
+*Structure* types define aggregates with heterogeneous fields that are _statically indexed_:
+```
+(type $time (struct (field i32) (field f64)))
+(type $point (struct (field $x f64) (field $y f64) (field $z f64)))
+```
+Such types can be used by forming *reference types*, which are a new type of value type:
+```
+(local $var (ref $point))
+```
+
+Fields are *accessed* with generic load/store instructions that take a reference to a structure:
+```
+(func $f (param $p (ref $point))
+  (store_field $point $y (get_local $p)
+    (load_field $point $x (get_local $p))
+  )
+)
+```
+All accesses are type-checked at validation time.
+
+Structures are *allocated* witha  `new` instruction that take initialization values for each field.
+The operator yields a reference to the respective type:
+```
+(func $g
+  (call $g (new_struct $point (i32.const 1) (i32.const 2) (i32.const 3)))
+)
+```
+Structures are garbage-collected.
+
+Structures can be compared for identity:
+```
+(same (new $point ...) (new $point ...))  ;; false
+```
+TODO: Could even allow heterogeneous equality (equality between operands of different type), but that might lead to some discontinueties or even prevent some potential optimizations?
+
+
+### Arrays
+
+*Array* types define aggregates with _homogeneous elements_ that are _dynamically indexed_:
+```
+(type $vector (array f64))
+(type $matrix (array (type $vector)))
+```
+Such types again can be used by forming reference types.
+For now, we assume that all array types have dynamic ([flexible](#flexible-aggregates)) size.
+
+Elements are accessed with generic load/store instructions that take a reference to an array:
+```
+(func $f (param $v (ref $vector))
+  (store_elem $vector (get_local $v) (i32.const 1)
+    (load_elem $vector (get_local $v) (i32.const 2))
+  )
+)
+```
+The element type of every access is checked at validation time.
+The index is checked at execution time.
+A trap occurs if the index is out of bounds.
+
+Arrays are *allocated* with a `new` instruction that takes a size and an initialization value as operands, yielding a reference:
+```
+(func $g
+  (call $g (new_array $vector (i32.const 0) (i64.const 3.14)))
+)
+```
+Arrays are garbage-collected.
+
+The *length* of an array, i.e., the number of elements, can be inquired via the `load_length` instruction:
+```
+(load_length $vector (get_local $v))
+```
+
+Like structures, arrays can be compared for identity.
+
```

Comment:
In JVM, arrays are first-class objects that extend `java.lang.Object`, support `getClass()`, `hashCode`, etc. In my implementations arrays are actually objects that have corresponding header, which points to class with vtable. How I can implement this with Wasm arrays? The only approach I can see is to generate additional wrapper class (two structures, one for data and one for vtable).